include/swift/AST/DiagnosticEngine.h

```diff
@@ -629,6 +634,42 @@ namespace swift {
     DiagnosticState(DiagnosticState &&) = default;
     DiagnosticState &operator=(DiagnosticState &&) = default;
   };
+
+  /// Abstract interface class for providing diagnostics from YAML files.
+  class YAMLDiagnostic {
+    virtual void getDiagnostics(std::string locale, std::string path) = 0;
+  public:
+    virtual ~YAMLDiagnostic() { }
+  };
+
+  class YAMLDiagnosticProvider : public YAMLDiagnostic {
+    /// Diagnostic messages language code.
+    std::string diagnosticLocaleCode = "";
+    
+    /// Diagnostic messages directory path.
+    std::string diagnosticMessagesPath = "";
```

Comment:
Is this every going to be used anywhere besides constructor of this producer?

Reply:
I don't think so, currently. I just thought that they might come in handy later!? If at any point we needed this info about `locale code` or `diag path` we will find them.

Reply:
That could be always added later than, when it's needed.

Reply:
One more thing I forgot to mention - I think we shouldn't require `path` to be always present it could be inferred based on locale and default directory path we have in CMakeList.txt right?

Reply:
I actually set the `DiagnosticMessagesPath` in CompilerInvocations as you can [see here at L69](https://github.com/HassanElDesouky/swift/blob/23934fd6e3dc8145b997a694b8241338dc6575e3/lib/Frontend/CompilerInvocation.cpp#L69)

If we used the `-diagnostic-messages-path` flag, the `DiagnosticMessagesPath` gets overridden. 

Therefore, I think we should require `path` to be always present. What do you think?

Reply:
Sure, makes sense!